docs/abi-spec.rst

```diff
@@ -0,0 +1,327 @@
+
+- `uint<M>`: `enc(X)` is the big-endian encoding of `X`, padded on the higher-order (left) side with zero-bytes such that the length is a multiple of 32 bytes.
+- `address`: as in the `uint160` case
+- `int<M>`: `enc(X)` is the big-endian two's complement encoding of `X`, padded on the higher-oder (left) side with `0xff` for negative `X` and with zero bytes for positive `X` such that the length is a multiple of 32 bytes.
+- `bool`: as in the `uint8` case, where `1` is used for `true` and `0` for `false`
+- `fixed<M>x<N>`: `enc(X)` is `enc(X * 2**N)` where `X * 2**N` is interpreted as a `int256`.
+- `fixed`: as in the `fixed128x128` case
+- `ufixed<M>x<N>`: `enc(X)` is `enc(X * 2**N)` where `X * 2**N` is interpreted as a `uint256`.
+- `ufixed`: as in the `ufixed128x128` case
+- `bytes<M>`: `enc(X)` is the sequence of bytes in `X` padded with zero-bytes to a length of 32.
+
+Note that for any `X`, `len(enc(X))` is a multiple of 32.
+
+## Function Selector and Argument Encoding
+
+All in all, a call to the function `f` with parameters `a_1, ..., a_n` is encoded as
+
+  `function_selector(f) enc([a_1, ..., a_n])`
+
+and the return values `v_1, ..., v_k` of `f` are encoded as
+
+  `enc([v_1, ..., v_k])`
+
+where the types of `[a_1, ..., a_n]` and `[v_1, ..., v_k]` are assumed to be
+fixed-size arrays of length `n` and `k`, respectively. Note that strictly,
+`[a_1, ..., a_n]` can be an "array" with elements of different types, but the
+encoding is still well-defined as the assumed common type `T` (above) is not
+actually used.
+
+## Examples
+
+Given the contract:
+
+```js
+contract Foo {
+  function bar(fixed[2] xy) {}
+  function baz(uint32 x, bool y) returns (bool r) { r = x > 32 || y; }
+  function sam(bytes name, bool z, uint[] data) {}
+}
+```
+
+Thus for our `Foo` example if we wanted to call `baz` with the parameters `69` and `true`, we would pass 68 bytes total, which can be broken down into:
+
+- `0xcdcd77c0`: the Method ID. This is derived as the first 4 bytes of the Keccak hash of the ASCII form of the signature `baz(uint32,bool)`.
+- `0x0000000000000000000000000000000000000000000000000000000000000045`: the first parameter, a uint32 value `69` padded to 32 bytes
+- `0x0000000000000000000000000000000000000000000000000000000000000001`: the second parameter - boolean `true`, padded to 32 bytes
+
+In total:
+```
+0xcdcd77c000000000000000000000000000000000000000000000000000000000000000450000000000000000000000000000000000000000000000000000000000000001
+```
+It returns a single `bool`. If, for example, it were to return `false`, its output would be the single byte array `0x0000000000000000000000000000000000000000000000000000000000000000`, a single bool.
+
+If we wanted to call `bar` with the argument `[2.125, 8.5]`, we would pass 68 bytes total, broken down into:
```

Comment:
This whole section needs rewriting as fixed points are now going to be decimals.